src/psygnal/_group.py

```diff
@@ -250,54 +254,58 @@ class MySignals(SignalGroup):
     ```
     """
 
-    _psygnal_signals: ClassVar[Mapping[str, Signal]]
-    _psygnal_instances: dict[str, SignalInstance]
+    _psygnal_signals: ClassVar[Mapping[str, Signal]] = {}
     _psygnal_uniform: ClassVar[bool] = False
 
+    _psygnal_instances: dict[str, SignalInstance]
+    _psygnal_relay: SignalRelay
+
     def __init__(self, instance: Any = None) -> None:
         cls = type(self)
         if not hasattr(cls, "_psygnal_signals"):  # pragma: no cover
             raise TypeError(
                 "Cannot instantiate `SignalGroup` directly.  Use a subclass instead."
             )
 
+        # Attach SignalInstance to this SignalGroup instance
         self._psygnal_instances = {
-            name: sig.__get__(self, cls) for name, sig in cls._psygnal_signals.items()
+            name: sig._get_signal_instance(self)
+            for name, sig in cls._psygnal_signals.items()
         }
+        # Attach SignalRelay to the object instance
         self._psygnal_relay = SignalRelay(self._psygnal_instances, instance)
 
     def __init_subclass__(cls, strict: bool = False) -> None:
         """Collects all Signal instances on the class under `cls._psygnal_signals`."""
-        cls._psygnal_signals = {
-            k: val
-            for k, val in getattr(cls, "__dict__", {}).items()
-            if isinstance(val, Signal)
+        # Collect Signals and remove from class attributes
+        # Use dir(cls) instead of cls.__dict__ to get attributes from super()
+        _psygnal_signals = {}
+        for k in dir(cls):
+            val = getattr(cls, k, None)
+            if isinstance(val, Signal):
+                _psygnal_signals[k] = val
+                delattr(cls, k)
```

Comment:
This comment is half note-to-self (explaining something i've just understood) and half note to @getzze  :)

I dislike this particular `delattr(cls, k)` here.  It leads to the rather surprising behavior:

```python
class MyGroup(SignalGroup):
    my_signal = Signal()

hasattr(MyGroup, 'my_signal')  # False
```
Any external observer would be absolutely justified to be rather surprised by that. And I'd like to avoid surprising behavior.

However, I just spent the last couple hour or two to make it work without that, and then I came into the same challenge you presumably encountered with the fact that we (by default) cache `SignalInstances` on the object itself, similar to how `functools.cached_property` works.  That is

```python
my_group()
# at this first access of `my_signal` on `my_group`, the newly created SignalInstance
# is set as an attribute on the my_group instance, overriding the `Signal` on the class
my_group.my_signal 
```

I see now why this PR has to also modify the caching behavior of `Signal.__get__`.  It's because of this one `delattr(cls, k)` line here:  by removing the `Signal` descriptor object, we're free to do our own behavior in `getattr` (which is desire able for controlling name conflicts), but then of course we want to control the caching of that object as well.

I also completely recognize that all of this ugliness stems from the original design of the `SignalGroup` API design itself.  Specifically: while the `Signal` descriptor protocol may works nicely as a class attribute on a regular class, it doesn't work quite as well on `SignalGroup` (inasmuch as we also want that group to have lots of other possible names on it, like `connect`, etc..)

*side note: part of the motivation for the descriptor protocol is that it makes type checking and IDE support much easier.  One thing I dislike about most observer-pattern implementations, is that you never know if you're calling the callback properly... and the descriptor pattern helps us to hint at that during static type checking (even if we haven't yet taken full advantage of that)*

Anyway, after banging my head against this for a little while, I'm again torn as to whether this is a good idea.  We either ask everyone to just use `group.all` (and end up with a clean pattern), or we provide the magical convenience of passing through all of these SignalRelay names, but possibly at the cost of rather surprising behavior for anyone unlucky enough to come into a conflict).  

still thinking...